main/src/library/LazyList.flix

```diff
@@ -0,0 +1,111 @@
+    /// The empty list is always strict.
+    ///
+    pub enum LazyList[t] {
+        case Empty,
+        case Cons(t, Lazy[LazyList[t]])
+    }
+
+    pub def empty(): LazyList[a] = Empty
+
+    pub def isEmpty(xs: LazyList[a]): Bool = xs == Empty
+
+    pub def cons(x: a, xs: LazyList[a]): LazyList[a] =
+        Cons(x, lazy xs)
+
+    pub def repeat(x: a): LazyList[a] =
+        unfold(_ -> Some(x, x), x)
+
+    pub def head(xs: LazyList[a]): Option[a] = match xs {
+        case Empty => None
+        case Cons(h, _) => Some(h)
+    }
+
+///    TODO: Unable to unify the types: 'Lazy' and 'LazyList'.
+///    This seems to be caused by the compiler transforming
+///    x :: acc to Cons(x, acc) which calls the Cons constructor in
+///    LazyList instead of List
+///
+///    pub def toList(xs: LazyList[a]): List[a] =
+///        foldRight((x, acc) -> x :: acc, Nil, xs)
+
+    pub def append(xs: LazyList[a], ys: LazyList[a]): LazyList[a] = match xs {
+        case Empty => ys
+        case Cons(h, t) => cons(h, append(force t, ys))
+    }
+
+    pub def fold(f: (b, a) -> b, s: b, xs: LazyList[a]): b = foldLeft(f, s, xs)
+
+    pub def foldLeft(f: (b, a) -> b, s: b, xs: LazyList[a]): b = match xs {
+        case Empty => s
+        case Cons(h, t) => foldLeft(f, f(s, h), force t)
+    }
+
+    /// TODO: How can we support early termination?
+    pub def foldRight(f: (a, b) -> b, s: b, xs: LazyList[a]): b = match xs {
+        case Empty => s
+        case Cons(h, t) => f(h, foldRight(f, s, force t))
+    }
+
+    pub def exists(f: a -> Bool, xs: LazyList[a]): Bool = match xs {
+        case Empty => false
+        case Cons(h, t) => f(h) || exists(f, force t)
+    }
+
+    pub def forall(f: a -> Bool, xs: LazyList[a]): Bool = match xs {
+        case Empty => true
+        case Cons(h, t) => f(h) && forall(f, force t)
+    }
+
+    pub def map(f: a -> b, xs: LazyList[a]): LazyList[b] = match xs {
+        case Empty => Empty
+        case Cons(h, t) => cons(f(h), map(f, force t))
+    }
+
+    pub def unfold(f: b -> Option[(a, b)], s: b): LazyList[a] = match f(s) {
+        case None => Empty
+        case Some((h, t)) => cons(h, unfold(f, t))
+    }
+
+    pub def zipWith(f: (a, b) -> c, xs: LazyList[a], ys: LazyList[b]): LazyList[c] =
+        zipWithHelper(f, xs, ys, Empty)
+
+    def zipWithHelper(f: (a, b) -> c, xs: LazyList[a], ys: LazyList[b], acc: LazyList[c]): LazyList[c] =
+        match (xs, ys) {
+            case (Empty, _) => Empty
+            case (_, Empty) => Empty
+            case (Cons(h1, t1), Cons(h2, t2)) => zipWithHelper(f, force t1, force t2, cons(f(h1, h2), acc))
+        }
+
+    pub def zip(xs: LazyList[a], ys: LazyList[b]): LazyList[(a, b)] = zipWith((x, y) -> (x, y), xs, ys)
+
+    pub def __eq(xs: LazyList[a], ys: LazyList[a]): Bool = match (xs, ys) {
+        case (Empty, Empty) => true
+        case (Cons(h1, t1), Cons(h2, t2)) => (h1 == h2) && __eq(force t1, force t2)
+        case _ => false
+    }
+
```

Comment:
A few more operations I'd like to see:
- drop
- take
- filter
- head
- last (!)
- reverse
- unfold/unfoldWithiter/unfoldWithOk

Reply:
I think I'm currently unable to implement `unfoldWithIter` and `unfoldWithOkIter` for LazyList2 as they are impure.
```
>> Unable to unify the Boolean formulas: 'true' and 'false'.

122 |     pub def unfoldWithIter(next: () ~> Option[a]): LazyList2[a] & Impure = LazyList2(lazy LazyList.unfoldWithIter(next))
                                                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                                                                           mismatched boolean formulas.
```

Reply:
Ah yes, I guess we have to leave those for now.